src/transformers/deepspeed.py

```diff
@@ -283,36 +160,30 @@ def trainer_config_finalize(self, args, model, num_training_steps):
             )
 
 
-# keep the config object global to be able to access it anywhere during TrainingArguments life-cycle
-_hf_deepspeed_config_weak_ref = None
-
-
-def set_hf_deepspeed_config(hf_deepspeed_config_obj):
-    # this is a special weakref global object to allow us to get to Deepspeed config from APIs
-    # that don't have an easy way to get to the Deepspeed config outside of the Trainer domain.
-    global _hf_deepspeed_config_weak_ref
-    # will go away automatically when HfDeepSpeedConfig is destroyed (when TrainingArguments is destroyed)
-    _hf_deepspeed_config_weak_ref = weakref.ref(hf_deepspeed_config_obj)
-
-
-def unset_hf_deepspeed_config():
-    # useful for unit tests to ensure the global state doesn't leak - call from `tearDown` method
-    global _hf_deepspeed_config_weak_ref
-    _hf_deepspeed_config_weak_ref = None
-
-
 def is_deepspeed_zero3_enabled():
-    if _hf_deepspeed_config_weak_ref is not None and _hf_deepspeed_config_weak_ref() is not None:
-        return _hf_deepspeed_config_weak_ref().is_zero3()
-    else:
-        return False
+    if (
+        "deepspeed_plugin" in AcceleratorState._shared_state
+        and AcceleratorState._shared_state["deepspeed_plugin"] is not None
+    ):
+        deepspeed_plugin = AcceleratorState._shared_state["deepspeed_plugin"]
+        return deepspeed_plugin.hf_ds_config.is_zero3() and deepspeed_plugin.zero3_init_flag
+    return False
 
 
 def deepspeed_config():
-    if _hf_deepspeed_config_weak_ref is not None and _hf_deepspeed_config_weak_ref() is not None:
-        return _hf_deepspeed_config_weak_ref().config
-    else:
-        return None
+    ds_config = None
+    if (
+        "deepspeed_plugin" in AcceleratorState._shared_state
+        and AcceleratorState._shared_state["deepspeed_plugin"] is not None
+    ):
+        ds_config = deepcopy(AcceleratorState._shared_state["deepspeed_plugin"].hf_ds_config.config)
+        if ds_config["gradient_accumulation_steps"] == "auto":
+            ds_config["gradient_accumulation_steps"] = 1
+        if "train_micro_batch_size_per_gpu" not in ds_config or ds_config["train_micro_batch_size_per_gpu"] == "auto":
+            ds_config["train_micro_batch_size_per_gpu"] = 1
+        if ds_config["train_batch_size"] == "auto":
+            del ds_config["train_batch_size"]
```

Comment:
Why are we removing `autos` here? They could be resolved with the `TrainingArguments` no? Or ar those auto values that Accelerate resolves but not Transformers?

Reply:
In `Accelerate` autos are handled in `prepare` method call but we will be calling `AutoModel.from_pretrained()` or `AutoModel.from_config()` way before it and as such few compulsory fields will be `auto` in accelerate setup when checking zero init is required in stage 3. So, this handles those situations in accelerate and is needed else deepspeed will throw errors when these fields are not correctly populated. 

Transformers Trainer has no issues as all the important population of `auto` fields already happen in `training_args.py`.

This is needed for accelerate integration to work properly when we want to use zero init in stage 3 for directly sharding parameters to GPUs.